### PR TITLE
feat: web bookmarks for arbitrary URLs (NIP-B0)

### LINKED ISSUE #25: Feature request: web bookmarks (NIP-B0)
## Summary

Split out of #21. That issue is now scoped to in-app bookmarks (NIP-51, kinds `10003`/`30003` — saving notes/articles). This issue covers the other half of the original feedback: bookmarking arbitrary **web pages**, per [NIP-B0](https://github.com/nostr-protocol/nips/blob/master/B0.md).

## NIP-B0 — Web Bookmarking

Kind `39701` — an addressable event *per URL*:

- `d` tag is the URI (scheme/host stripped for `https`, e.g. `alice.blog/post`), so bookmarks of the same page are addressable/editable rather than duplicated.
- `.content` is a free-text description of the bookmark (may be empty).
- Optional tags: `title` (usable as the HTML link's title/attribute), `t` (topic hashtags), `published_at` (unix seconds, first time the bookmark was published).
- Comments on a web bookmark use NIP-22 (kind `1111`), not raw kind-1 replies.

Example:

```json
{
  "kind": 39701,
  "tags": [
    ["d", "alice.blog/post"],
    ["published_at", "1738863000"],
    ["title", "Blog insights by Alice"],
    ["t", "post"],
    ["t", "insight"]
  ],
  "content": "A marvelous insight by Alice about the nature of blogs and posts."
}
```

## Scope

- A way to save a bookmark for an arbitrary URL (title/description/tags), publishing a kind `39701` event.
- A view listing the current user's web bookmarks: `{ kinds: [39701], authors: [pubkey] }`.
- Distinct from #21's in-app bookmark list — this is one event per URL, not an entry appended to a list.

See also #21 (in-app note/article bookmarks, NIP-51).

### PR DESCRIPTION
**Pull request by mroxso**
## Summary

Implements NIP-B0 web bookmarks (kind `39701`) — split out from #21 since it's a different event shape: one addressable event per URL (with its own title/description/tags) rather than an entry in a NIP-51 list.

- `src/hooks/useWebBookmarks.ts`: `useMyWebBookmarks`, `useCreateWebBookmark`, `useDeleteWebBookmark`, and `bookmarkDTag`/`bookmarkUrl` implementing the spec's "strip `https://` for the `d` tag" rule (and reversing it for display/linking).
- Delete sends a NIP-09 kind `5` deletion request, and also removes the item from the local query cache directly — relays aren't obligated to honor deletion requests, so a plain refetch could still hand it back.
- `src/apps/web-bookmarks/index.tsx`: new app, registered in `src/os/registry.ts`. Inline "Add bookmark" form (URL, optional title/description/tags), list view with an external-link-out to the saved URL, tag badges, and a hover "Remove" action.
- `docs/apps.md` updated.

## Test plan
- [x] `npx tsc --noEmit`
- [x] `npx eslint --cache`
- [x] `npx vitest run src/hooks/useWebBookmarks.test.ts` — `bookmarkDTag`/`bookmarkUrl` round-trip for both `https://` and non-`https` schemes
- [x] Drove the running app with Playwright: logged in with a locally generated nsec, added a bookmark (URL + title + description + tags), confirmed the publish succeeded and a toast confirmed it, reloaded, and confirmed the Web Bookmarks app lists it with the stripped `d` tag, description and tag badges intact and opens the URL in a new tab.
- [x] Confirmed the signed-out state shows the sign-in prompt.

Closes #25

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01BYiUtZMQeA5RHggQw73wto

### CODE REVIEW

File: src/apps/web-bookmarks/index.tsx [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+  );
+}
+
+function WebBookmarkRow({ event }: { event: NostrEvent }) {
+  const del = useDeleteWebBookmark();
+  const { toast } = useToast();
+
+  const dTag = tagValue(event, 'd') ?? '';
+  const url = sanitizeUrl(bookmarkUrl(dTag)) ?? bookmarkUrl(dTag);
+  const title = tagValue(event, 'title') ?? dTag;
```
[the remark this answers is not in the document]
Comment (mroxso) [reply src/apps/web-bookmarks/index.tsx:170]: Fixed — the link only renders when sanitizeUrl() succeeds; otherwise it's plain text with a title explaining why, instead of falling back to the unsanitized value.

File: src/apps/web-bookmarks/index.tsx [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+  const [url, setUrl] = useState('');
+  const [title, setTitle] = useState('');
+  const [description, setDescription] = useState('');
+  const [tags, setTags] = useState('');
+  const create = useCreateWebBookmark();
+  const { toast } = useToast();
+
+  const trimmedUrl = url.trim();
+  const isValid = /^https?:\/\/.+/i.test(trimmedUrl);
+
```
[the remark this answers is not in the document]
Comment (mroxso) [reply src/apps/web-bookmarks/index.tsx:104]: Fixed — added a dedicated isBookmarkableUrl() check that allows the same schemes sanitizeUrl() does (https, http, mailto, nostr). Deliberately not reusing sanitizeUrl() itself for validation, since it resolves relative URLs against window.location.origin and would incorrectly 'validate' a bare hostname like example.com as a link back into this app.

File: src/hooks/useWebBookmarks.ts [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+  return ['nostr', 'web-bookmarks', pubkey ?? ''] as const;
+}
+
+/**
+ * The `d` tag per NIP-B0: the URI with the `https://` scheme stripped (every
+ * other scheme keeps its full form, so it round-trips through `bookmarkUrl`).
+ */
+export function bookmarkDTag(url: string): string {
+  return url.startsWith('https://') ? url.slice('https://'.length) : url;
+}
```
[the remark this answers is not in the document]
Comment (mroxso) [reply src/hooks/useWebBookmarks.ts:25]: Fixed — the scheme check is now case-insensitive, with a regression test for HTTPS/HtTpS producing the same d tag as https.

File: src/hooks/useWebBookmarks.ts [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+ * to the same `d` tag instead of creating duplicate bookmarks.
+ */
+export function bookmarkDTag(url: string): string {
+  return HTTPS_SCHEME_RE.test(url) ? url.slice('https://'.length) : url;
+}
+
+/** Reconstructs a clickable URL from a `d` tag written by `bookmarkDTag`. */
+export function bookmarkUrl(dTag: string): string {
+  return /^[a-z][a-z0-9+.-]*:\/\//i.test(dTag) ? dTag : `https://${dTag}`;
+}
```
[the remark this answers is not in the document]
Comment (mroxso) [reply src/hooks/useWebBookmarks.ts:38]: Fixed — mailto:/nostr: are now recognized via a closed list of the opaque schemes this app supports, alongside the existing '://' check for hierarchical schemes (kept so gemini:// etc. still round-trip, and so a stripped https URL with a port isn't misread as a scheme). Verified live: a mailto: bookmark now round-trips to a working mailto: link after reload.

File: src/hooks/useWebBookmarks.ts [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    mutationFn: async ({ url, title, description, tags }: WebBookmarkInput) => {
+      if (!user) throw new Error('Sign in to bookmark a page');
+
+      const eventTags: string[][] = [['d', bookmarkDTag(url)]];
+      if (title?.trim()) eventTags.push(['title', title.trim()]);
+      for (const tag of tags ?? []) {
+        if (tag.trim()) eventTags.push(['t', tag.trim().toLowerCase()]);
+      }
+      eventTags.push(['published_at', String(Math.floor(Date.now() / 1000))]);
+
```
[the remark this answers is not in the document]
Comment (mroxso) [reply src/hooks/useWebBookmarks.ts]: Fixed — the create mutation now looks up the existing bookmark for the same d tag before publishing and carries its published_at forward instead of resetting it to now.

File: src/hooks/useWebBookmarks.test.ts [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+  it('round-trips an https URL back through bookmarkUrl', () => {
+    const url = 'https://alice.blog/post';
+    expect(bookmarkUrl(bookmarkDTag(url))).toBe(url);
+  });
+
+  it('round-trips a non-https URL back through bookmarkUrl', () => {
+    const url = 'http://alice.blog/post';
+    expect(bookmarkUrl(bookmarkDTag(url))).toBe(url);
+  });
```
[the remark this answers is not in the document]
Comment (mroxso) [reply src/hooks/useWebBookmarks.test.ts:27]: Added — tests for mailto:/nostr: round-tripping, a hierarchical non-https scheme (gemini://), and the port-number edge case (alice.blog:8080/post correctly gets https:// prepended, not misread as having a scheme).

**Comment by mroxso** [comment]
Addressed the remaining "previously missed" finding from the earlier review pass:

- `useMyWebBookmarks()` now dedupes multiple revisions of the same addressable event (same `d` tag) instead of showing duplicate rows — extracted as a tested `dedupeLatestByDTag()` helper that keeps the newest per `d`.

(The `published_at` semantics finding from that same pass was already fixed in an earlier commit on this PR.)

🤖 Generated with [Claude Code](https://claude.com/claude-code)
